### PR TITLE
feat(api): add support for lua function & description in keymap

### LINKED ISSUE #6123: Descriptive Maps
I was thinking about trying to add some introspection to mappings built into nvim. I have seen a few plugins out there, and worked on one myself, that would allow you to inspect your mappings with some information in them. Most of them rely on adding a plugin, so I don't see widespread adoption of them, and some of them are slow because they're implemented in vimscript, as opposed to being builtin parts of the maps.

Seeing some presentations of spacemacs, with the abiilty to see what a mapping means (in human words) in real time, inspired me to think about how to get something done like that in vim. I think other plugin maintainers could certainly use something like this, for example a host could be made for denite or similar plugins that would get all your mappings and your descriptions, and you could search through them. It would also be helpful for people when they get a new plugin, they could see what mappings that plugin has without have to dig through all the docs again. It would also allow people to have more "living documentation" where the comments they write in the code show up in front of the user, rather than having to keep two locations for information (where you write the code and the `docs/` folder).

I was thinking about trying to add some stuff into nvim so people could do something like this:

```vim
" Not sure I like one more than the other here...

" Add a new modifier to a map, with a "end" modifier, where a description can be added
nnoremap <describe> lhs rhs
    \ <description> Now I can type whatever I want to describe this.

" New command where you specify the mode, the lhs and then a description.
nnoremap lhs rhs
if has('nvim-version')
    describe n lhs Now I can type whatever I want to describe this
endif
```

Then you would be able to get them with the `maparg` command:

```vim

let map_dict = maparg('lhs', 'n', v:false, v:true)
echo map_dict
" { ..., 'description': 'Now I can type whatever I want to describe this'}
```

I guess a couple questions:
    - Does anyone have any other suggestions, better ideas for this concept?
    - If implemented, would this be accepted?
    - Does anyoen have any idea if this would get accepted at Vim?

Anyways, I'm open to suggestions or rejections :)

### LINKED ISSUE #14976: Allow `vim.api.nvim_set_keymap` to accept a lua callback function
My apologies if this already exists or has already been requested. I did try searching for it, but couldn't come up with any keywords to bring up any related issues.

`vim.api.nvim_set_keymap` currently takes as its third argument a string, which is a vim expression that will be called when the mapping is activated. This can be used to call a lua function by running `v:lua.SomeGlobalFunction` or with `<cmd>lua require('file_that_exports_function').some_exported_function()<cr>`.

My request would be to allow the third argument to optionally be a lua function, which will be called when the mapping is activated.

Currently legal example:
```lua
function TabComplete()
  -- implementation irrelevant to example
end

vim.api.nvim_set_keymap("i", "<Tab>", "v:lua.TabComplete()", {expr = true})
```

Here, `TabComplete` must be a global function, so that it can be called as a vimscript expression. This means I must ensure to name it such that it hopefully won't collide with any other globally named functions defined by other plugins.

Desired example:
```lua
local function tab_complete()
  -- implementation irrelevant to example
end

vim.api.nvim_set_keymap("i", "<Tab>", function() tab_complete() end, {expr = true})
```

Here we solve the problem of cluttering the global namespace. Additionally, if it were a short function, we could inline it directly in the keymap call. And any other local values defined within the scope will be accessible to the callback without needing to be made global.

Is this something that is currently possible, or may be considered for a future release? As shown with my legal example, this is not something that is currently blocking me. However it would be a preferrable syntax.

### PR DESCRIPTION
**Pull request by shadmansaleh**
### Added support for lua function in keymaps in
--------------------------------------------
- nvim_set_keymap
  Can set lua function as keymap rhs like following:
```lua
vim.api.nvim_{buf_}set_keymap('n', '<leader>lr', '', {callback = vim.lsp.buf.references})
```
  Note: lua function can only be set from lua . If api function being
  called from viml or over rpc this option isn't available.
- nvim_{buf_}get_keymap
  When called from lua, lua function is returned is `callback` key .
  But in other cases callback contains number of the function ref.
- :umap, nvim_del_keymap & nvim_buf_del_keymap clears lua keymaps correctly.
- :map commands for displaing rhs .
   For lua keymaps rhs is displayed as <Lua function ref_no>
   Note: lua keymap cannot be set through viml command / functions.
- mapargs()
  When dict is false it returns string in `<Lua function ref_no>`
  format (same format as :map commands).
  When dict is true it returns ref_no number in `callback` key.
- mapcheck()
  returns string in `<Lua function ref_no>` format (same format as :map commands).

### Added support for keymap description
---------------------------------------
- nvim_{buf_}set_keymap: added `desc` option in opts table .
 ```lua
vim.api.nvim_set_keymap('n', '<leader>w', '<cmd>w<cr>', {desc='Save current file'})
```
- nvim_{buf_}get_keymap: contains `desc` in returned list.
- commands like `:nmap <leader>w` will show description in a new line below rhs.
- `maparg()` return dict contains `desc`.

closes #14976
closes #6123

### CODE REVIEW

File: src/nvim/api/private/helpers.c
Comment (bfredl) [inline src/nvim/api/private/helpers.c:608]: no need to check `channel_id` , only a lua caller can produce a `kObjectTypeLuaRef` (and if there was a way, it wouldn't be a problem either..)

Comment (shadmansaleh) [reply src/nvim/api/private/helpers.c:608]: Ok I'll remove those.

**Comment by numToStr** [comment]
> * nvim_set_keymaps
> * nvim_buf_set_keymaps
>
> * nvim_get_keymaps
> * nvim_buf_get_keymaps

@shadmansaleh I suppose these should be singular i.e `nvim_set_keymap`?

**Comment by shadmansaleh** [comment]
> @shadmansaleh I suppose these should be singular i.e nvim_set_keymap?

These are already existing api functions. I'm just extending their functionality .

**Comment by numToStr** [comment]
I was talking about the typo in the description.

File: src/nvim/api/private/helpers.c
Comment (gpanders) [reply src/nvim/api/private/helpers.c:608]: >And this is a helper function for api not an api function, I'm not sure what you meant by it will always be defined per the API contract.

Yep, you're right! This was the part I missed in my review :)

File: src/nvim/getchar.c
Comment (shadmansaleh) [reply src/nvim/getchar.c:3974]: Shouldn't we through error if mapped function causes an error ? It's not implemented yet . But I do want to through error here.

Comment (gpanders) [reply src/nvim/getchar.c:3974]: Yes I think so too.

File: src/nvim/getchar.c
Comment (shadmansaleh) [inline src/nvim/getchar.c:3969]: Should we apply `replace_termcodes` on result of function ?

File: src/nvim/api/private/helpers.c
Comment (shadmansaleh) [inline src/nvim/api/private/helpers.c:1092]: This doesn't work if I try to put anything in the dict (I tried even putting a bool) I get a `string length overflow` error from `luajit` . Any idea why ?

File: src/nvim/getchar.c
Comment (bfredl) [reply src/nvim/getchar.c:3971]: Too magic. I'd rather add a shortcut for `vim.api.nvim_replace_termcodes(str, true, true, true)` (like `vim.termcode(str)` or something ) in the next PR.

**Comment by shadmansaleh** [comment]
Waiting for review . I think I've got everything related to keymaps (If there's somthing left let me know). See https://github.com/neovim/neovim/pull/16594#issue-798921436 for summary on behavioral changes .

**Comment by shadmansaleh** [comment]
> The luaref function executes in textlock mode, so many things someone might like to do require using vim.schedule

Regular expr mappings are ran under textlock that's why lua functions are also ran under textlock . I assume there's a reason why exprs are ran under textlock and running without textlock may cause unexpected issues . Though in this case others will know better then me .

> I do think we should add a note to the docs explaining that the luaref function is called with textlock active.

Indeed.

> Even if I use vim.schedule in the above example, I still have to press another key after entering ]g for the mapping to execute

You're right. It seems like something is blocking after function is ran. I need to read more of how keymaps are handled. If any one has a clue on what might be blocking please let me know 😄

> letting the rhs argument be nil when luaref is present

The internal code doesn't care whether rhs is nil or '' as for lua keymaps NULL is stored for rhs . What's blocking is lua api bindings . Those are generated from c function definitions . I'm not completely sure how to make them accept multiple types  🤔

**Comment by zeertzjq** [comment]
I think `<expr>` mapping are run under textlock because it isn't expected for `<expr>` to have side-effects. `<expr>` is just used for evaluation, not for actual action. Meanwhile Lua mappings should be more similar to `<Cmd>` mappings.

**Comment by zeertzjq** [comment]
I feel for non-`<expr>` LuaRef mappings, there can be an alternative implementation: encode a LuaRef as a keycode and insert it into the typeahead buffer. It can be later handled like `K_COMMAND`.

**Comment by shadmansaleh** [comment]
I think we can allow non-expr lua keymaps to execute out of textlock . Also I'll take a look at how K_COMMAND works.

**Comment by bfredl** [comment]
Yes, lua mappings which are not `expr=true` need to use some mechanism similar to / based on `K_COMMAND`.

**Comment by shadmansaleh** [comment]
@bfredl is `K_COMMAND` implementation isolated in https://github.com/neovim/neovim/pull/4419 ?

**Comment by zeertzjq** [comment]
BTW, `K_COMMAND` only appears 15 times in source code.

**Comment by bfredl** [comment]
Yes, you find all places where `K_COMMAND` needs to be handled (one per interactive mode) by grepping for it :]

**Comment by seblj** [comment]
I have been following this PR, and the one for commands https://github.com/neovim/neovim/pull/16752

@bfredl suggested to have the luaref as part of the `command` in the other PR, instead og having it in the opts table. Personally I like the approach he suggested more than specifying it in the opts table. So I was just wondering if it is possible to do here as well?

**Comment by zeertzjq** [comment]
I think this will break backward-compatibility.

**Comment by lewis6991** [comment]
> I think this will break backward-compatibility.

How?

**Comment by zeertzjq** [comment]
> > I think this will break backward-compatibility.
> 
> How?

Oh, I didn't look at that thread carefully.

File: src/nvim/api/private/helpers.c
Comment (gpanders) [inline src/nvim/api/private/helpers.c:615]: I would recommend also adding a `desc` option for when `rhs` is a function (similar to #16752) that is listed when users use `:map {lhs}`.

Comment (shadmansaleh) [reply src/nvim/api/private/helpers.c:615]: Good idea . I was initially thinking of adding it on a followup . But can be added right here .

**Comment by shadmansaleh** [comment]
> Yes, lua mappings which are not expr=true need to use some mechanism similar to / based on K_COMMAND.

@bfredl why though ? In case of K_COMMAND I can see that those maps need to be executed in command mode and `:` might not guarantee same state of command mode I assume . But incase of lua why can't we directly call the lua callback in time of map evaluation ? What's the added benefit of K_COMMAND like structure in this case ?

**Comment by zeertzjq** [comment]
I think some operations may be not safe to do when the mapping is evaluated.

**Comment by bfredl** [comment]
@shadmansaleh K_COMMAND has nothing to with "command mode", on the contrary it executes a command safely according to the rules of each mappable mode. `<expr>` style execution has a lot of limitations like:
```
- Changing the buffer text |textlock|.
- Editing another buffer.
- The |:normal| command.
- Moving the cursor is allowed, but it is restored afterwards.
- If the cmdline is changed, the old text and cursor position are restored.
```
which it would be risky to try to disable as code calling into `getchar()` will not be prepared for them to happen.

**Comment by shadmansaleh** [comment]
In that case I misunderstood it . Thanks @zeertzjq @bfredl

File: src/nvim/api/vim.c
Comment (bfredl) [inline src/nvim/api/vim.c:1575]: Type of existing param cannot change in a stable API method, we'll need to use `opts` in this case.

File: src/nvim/getchar.c
Comment (bfredl) [inline src/nvim/getchar.c:4630]: `nlua` is not the scope, so it should be `map_execute_lua` or something, also needs `void` to make it a prototype.

**Comment by zeertzjq** [comment]
Oh, there are merge conflicts now.

File: src/nvim/api/vim.c
Comment (shadmansaleh) [reply src/nvim/api/vim.c:1575]: Isn't this still backward compatible ?

File: src/nvim/getchar.c
Comment (bfredl) [inline src/nvim/getchar.c:2649]: use `LUA_NOREF` directly as sentinel value: `if (rhs_lua == NO_REF) ...`

File: src/nvim/api/keysets.lua
Comment (bfredl) [inline src/nvim/api/keysets.lua:33]: These two look duplicate, we should settle for one. (normally we would use `on_xx` for callback args though I am not sure what xx would be..)

Comment (shadmansaleh) [reply src/nvim/api/keysets.lua:33]: `cb` is just alias for callback . like set_hl can have fg/bg as alias.

I'm not sure if `on_xx` is appropriate here .

File: src/nvim/getchar.c
Comment (gpanders) [inline src/nvim/getchar.c:3519]: Not sure about putting `desc` on a newline. If plugins or scripts depend on there being one mapping per line this would break that.

Knowing the number of the Lua ref is (imo) not that useful or interesting. What if instead we indicate a mapping uses a Lua function by prepending it with an `L` or `#` (similar to how buffer-local mappings have a `@` and non-remappable mappings use `*`). So e.g. `:map` would show

```
n  ZT          L Open a new tab
```

This should be pretty straightforward to add. The relevant code is at line 3486 in `getchar.c` and the docs under `|map-listing|`.

**Review by gpanders — CHANGES_REQUESTED** [review]

File: src/nvim/api/keysets.lua
Comment (bfredl) [reply src/nvim/api/keysets.lua:33]: > like set_hl can have fg/bg as alias.

This is the special case. It doesn't mean we should go around and add lot's of aliases to freshly designed API:s. Let's pick one or the other. We are expected to create an "ergonomic" layer above this for init.lua anyway.

Comment (shadmansaleh) [reply src/nvim/api/keysets.lua:33]: ok then . will stick to callback only

File: src/nvim/api/private/helpers.c
Comment (shadmansaleh) [reply src/nvim/api/private/helpers.c:609]: If new_ref is done here the ref can only be cleared only on fail as the ref will be attached to keymap on success . freeing it at end of this function will make keymap fail.

File: src/nvim/api/private/helpers.c
Comment (shadmansaleh) [reply src/nvim/api/private/helpers.c:1103]: No they won't leak . They are attached to keymap and will be cleaned when keymap gets removed or overwritten .

File: src/nvim/getchar.c
Comment (shadmansaleh) [reply src/nvim/getchar.c:18]: I'm not sure I understamd what you're saying her . where should free_object & clear_error happen ?

File: src/nvim/getchar.c
Comment (shadmansaleh) [reply src/nvim/getchar.c:4660]: vgetorpeek cannot be called out of getchar . exposing it to  lua/executor feels worse.

rename from nlua_execute_map to map_execute_lua was suggested by bfredl

File: src/nvim/getchar.c
Comment (shadmansaleh) [reply src/nvim/getchar.c:3519]: But non lua keymaps can also have description . And lua keymap without description has to be shown as ref no any way . description being on separate line personally seems nicer.

Also I can't see why someone would be parsing `:map` command output when `maparg()` exists ?

File: src/nvim/api/private/helpers.c
Comment (gpanders) [reply src/nvim/api/private/helpers.c:609]: Yes, you only call `NLUA_CLEAR_REF` in the `fail_and_free` label, which is only used on error. If an error doesn't occur the ref is not freed.

File: src/nvim/getchar.c
Comment (bfredl) [reply src/nvim/getchar.c:4660]: this uses global state of the mapping engine, so I think getchar.c is more appropriate as it is. The nlua part is already abstracted as `nlua_call_ref`

File: src/nvim/api/private/helpers.c
Comment (gpanders) [reply src/nvim/api/private/helpers.c:1103]: Yes but you are making a new copy of the ref with `api_new_luaref`. The ref attached to the keymap will be freed when the keymap is deleted, but in this function you are only listing existing mappings. So when are those new copies freed?

Basically, does this work if you remove `api_new_luaref` and just copy the integer value directly?

File: src/nvim/getchar.c
Comment (gpanders) [reply src/nvim/getchar.c:4660]: >vgetorpeek cannot be called out of getchar . exposing it to lua/executor feels worse.

Ok fair enough, but including `private/helpers.h` still doesn't seem right to me, judging by the "private" in the path.

File: src/nvim/getchar.c
Comment (gpanders) [reply src/nvim/getchar.c:3519]: >Also I can't see why someone would be parsing :map command output when maparg() & nvim_get_keymap exists ?

Just because it doesn't make sense doesn't mean people don't do it :) Though I'm not sure anyone actually is doing it, so maybe my concerns aren't valid.

But you're right, `desc` can be provided for non-Lua mappings as well, so this is probably fine.

File: src/nvim/api/private/helpers.c
Comment (shadmansaleh) [reply src/nvim/api/private/helpers.c:1103]: Ignore my earlier comment I thought this was on different part of the code.

The entire return object including all the luarefs of this function will be cleared at end of  `nvim_get_keymap` after after result is pushed to lua stack .

File: src/nvim/api/private/helpers.c
Comment (shadmansaleh) [reply src/nvim/api/private/helpers.c:609]: fail_and_free is always executed not on just fail . but we can workarround that so ref doesn't get cleared on success.

File: src/nvim/getchar.c
Comment (bfredl) [reply src/nvim/getchar.c:4660]: It's an historical artifact that should be removed now that we can mix API and non-API functions in the same file. until then, including `private/helpers.h` to use the functions there is fine.

File: src/nvim/getchar.c
Comment (shadmansaleh) [reply src/nvim/getchar.c:18]: same as https://github.com/neovim/neovim/pull/16594#discussion_r776391779 ?

Comment (gpanders) [reply src/nvim/getchar.c:18]: Yes, per bfredl's comment this is fine.

**Review by gpanders — APPROVED** [review]

File: src/nvim/getchar.c
Comment (bfredl) [inline src/nvim/getchar.c:2668]: use a static buffer, like `char tmp_buf[64]`.

File: src/nvim/getchar.c
Comment (bfredl) [inline src/nvim/getchar.c:3040]: `XFREE_CLEAR` already handles NULL so this if and the next should be removed.

File: src/nvim/getchar.c
Comment (bfredl) [inline src/nvim/getchar.c:3061]: `xstrdup` avoids two casts.

**Comment by shadmansaleh** [comment]
Is this good to be merged ?

File: src/nvim/getchar.c
Comment (shadmansaleh) [reply src/nvim/getchar.c:3991]: Yep this can be removed . I had forgotten about that.

File: src/nvim/eval/funcs.c
Comment (bfredl) [inline src/nvim/eval/funcs.c:6034]: should we use desc here if available? Maybe `<Lua function: %s>` with the desc instead of opaque int.

**Comment by bfredl** [comment]
LGTM after comments, also a squash would be good :]

File: src/nvim/eval/funcs.c
Comment (shadmansaleh) [reply src/nvim/eval/funcs.c:6034]: I have thought having same representation like `:map` command will be better . Also makes it consistent with when desc isn't given . the desc is provided when `dict` option is used on `maparg`

File: src/nvim/getchar.c
Comment (shadmansaleh) [reply src/nvim/getchar.c]: done

File: src/nvim/eval/funcs.c
Comment (bfredl) [reply src/nvim/eval/funcs.c:6034]: > I have thought having same representation like :map command will be better . 

I agree, and shouldn't `:map` use desc as well? also seems more helpful than showing opaque number to the user.

Comment (shadmansaleh) [reply src/nvim/eval/funcs.c:6034]: `:map` shows desc in 2nd line . Because non lua keymaps can also have desc I feel like it'll be weird if for lua desc apeared in place of rhs and for non lua it apeared somewhere else.

Currently it gets shown like this.
```
:map xx
lhs   flags   rhs
                    desc
```

Comment (bfredl) [reply src/nvim/eval/funcs.c:6034]: ok, let's keep it like it is for now. we can always develop it further until 0.7 anyway.

**Comment by clason** [comment]
@shadmansaleh I'm getting a weird error with the following mapping (from https://github.com/L3MON4D3/LuaSnip, but I don't see an issue from the plugin itself):
```lua
map( 'i', '<tab>', '',
{callback= function()
  if require"luasnip".expand_or_jumpable() then
    vim.schedule(require"luasnip".expand_or_jump())
  else
    return "<tab>"
  end
end,
expr = true })
```
If I type a snippet and try to expand it with `<tab>`, I get
```
E5108: Error executing lua: ...site/pack/packer/start/luasnip/lua/luasnip/util/util.lua:191: Failed to save undo information
stack traceback:
        [C]: in function 'nvim_buf_set_text'
        ...site/pack/packer/start/luasnip/lua/luasnip/util/util.lua:191: in function 'put'
        ...ite/pack/packer/start/luasnip/lua/luasnip/nodes/node.lua:55: in function 'put_initial'
        .../pack/packer/start/luasnip/lua/luasnip/nodes/snippet.lua:559: in function 'put_initial'
        .../pack/packer/start/luasnip/lua/luasnip/nodes/snippet.lua:366: in function 'trigger_expand'
        ...nvim/site/pack/packer/start/luasnip/lua/luasnip/init.lua:167: in function 'snip_expand'
        ...nvim/site/pack/packer/start/luasnip/lua/luasnip/init.lua:240: in function 'expand'
        ...nvim/site/pack/packer/start/luasnip/lua/luasnip/init.lua:287: in function 'expand_or_jump'
        /Users/clason/.config/nvim/plugin/mappings.lua:17: in function </Users/clason/.config/nvim/plugin/mappings.lua:15>
```
(the "non-expanding" case works fine).

**Comment by clason** [comment]
(For those only interested in getting the LuaSnip keymap to work, you can of course do
```lua
vim.keymap.set( 'i', '<tab>', function()
  if not require"luasnip".expand_or_jump() then
    vim.api.nvim_feedkeys(vim.api.nvim_replace_termcodes('<tab>', true, false, true), 'n', true)
  end
end)
```
instead; but I still would like to understand what's going on with the "naive" translation.)

**Comment by shadmansaleh** [comment]
@clason I think you're getting that error because `expr` keymaps are executed under `textlock`.

**Comment by clason** [comment]
Yep, that matches my observations and sounds like a reasonable explanation.

But why is the equivalent expr-mapping in Vimscript working then? Does `textlock` not apply there? If so, that's a bit of a regression in the new (or rather, Neovim) API?

**Comment by shadmansaleh** [comment]
textlock applies to viml version too .

> ``lua
>vim.keymap.set( 'i', '<tab>', 
>  'luasnip#expand_or_jumpable() ? "<Plug>luasnip-expand-or-jump" : "<tab>"',
>  { expr = true })
>```
Viml version that you've given doesn't do text edit under expr map


equivalent lua will probably be
```lua
vim.keymap.set( 'i', '<tab>', function()
  if require"luasnip".expand_or_jumpable() then
    return "<Plug>luasnip-expand-or-jump"
  else
    return "<tab>"
  end,
  { expr = true })
```

**Comment by zeertzjq** [comment]
> just inserts a literal <Plug>luasnip-expand-or-jump

This is because `noremap = false` is only automatically set when rhs begins with `<plug>`, and in this case it doesn't.

**Comment by clason** [comment]
OK, that makes sense; then that's just a gotcha that is good to know. Thanks for explaining!

**Pull request merged**
